### PR TITLE
8259214: MetaspaceClosure support for Arrays of MetaspaceObj

### PR DESCRIPTION
**Pull request by iklam**
Currently, `MetaspaceClosure::push` supports only the following variants:

```
MetaspaceClosure* it = ...;
Klass* o = ...; it->push(&o);
Array<int>* a1 = ...; it->push(&a1);
Array<Annotation*>* a2 = ...; it->push(&a2);
```

In Valhalla, support is needed for the following variant (Annotation is a subtype of MetaspaceObj):

```
Array<Annotation>* a3 = ...; it->push(&a3);
```

This change will allow CDS to make a copy of this array, as well as relocating all the pointers embedded in the elements of this array. See a test case in test_metaspaceClosure.cpp.

I also cleaned up the code (with help from @kimbarrett) to use SFINAE to dispatch from `MetaspaceClosure::push` to the different subtypes of `MetaspaceClosure::Ref`.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259214](https://bugs.openjdk.java.net/browse/JDK-8259214): MetaspaceClosure support for Arrays of MetaspaceObj

### Reviewers
 * [Frederic Parain](https://openjdk.java.net/census#fparain) (@fparain - Committer) ⚠️ Review applies to 84074148057f65e8b8e27ae0ddb7d64681ae480f
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to 84074148057f65e8b8e27ae0ddb7d64681ae480f

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1995/head:pull/1995`
`$ git checkout pull/1995`

### CODE REVIEW

**Review by fparain — APPROVED** [review]
Looks good to me.
Thank you for adding this feature.

Fred

File: src/hotspot/share/memory/metaspaceClosure.hpp [hunk 1]
```diff
@@ -182,65 +185,80 @@ class MetaspaceClosure {
...
 
     // all Arrays are read-only by default
     virtual bool is_read_only_by_default() const { return true; }
     virtual bool not_null()                const { return dereference() != NULL;  }
     virtual int size()                     const { return dereference()->size(); }
     virtual MetaspaceObj::Type msotype()   const { return MetaspaceObj::array_type(sizeof(T)); }
+  };
+
+  // OtherArrayRef -- iterate an instance of Array<T>, where T is NOT a subtype of MetaspaceObj.
+  // T can be a primitive type, since as int, or a structure. However, we do not scan
```
Comment (calvinccheung) [inline src/hotspot/share/memory/metaspaceClosure.hpp:209]: typo: s/since/such

File: src/hotspot/share/memory/metaspaceClosure.hpp [hunk 2]
```diff
@@ -288,31 +306,69 @@ class MetaspaceClosure {
...
+    push_impl(new REF_TYPE(mpp, w));
+  }
+
+public:
+  // When MetaspaceClosure::push(...) is called, pick the correct Ref subtype to handle it:
+  //
+  // MetaspaceClosure*      it = ...;
+  // Klass*                 o  = ...;  it->push(&o);     => MSORef
+  // Array<int>*            a1 = ...;  it->push(&a1);    => OtherArrayRef
+  // Array<Annotation>*     a2 = ...;  it->push(&a3);    => MSOArrayRef
```
Comment (calvinccheung) [inline src/hotspot/share/memory/metaspaceClosure.hpp:321]: &a3 should be &a2?

File: src/hotspot/share/memory/metaspaceClosure.hpp [hunk 3]
```diff
@@ -288,31 +306,69 @@ class MetaspaceClosure {
...
+  // Array<Klass*>*         a3 = ...;  it->push(&a2);    => MSOPointerArrayRef
```
Comment (calvinccheung) [inline src/hotspot/share/memory/metaspaceClosure.hpp:322]: &a2 should be &a3?

File: src/hotspot/share/memory/metaspaceClosure.hpp [hunk 4]
```diff
@@ -288,31 +306,69 @@ class MetaspaceClosure {
...
+  // Array<Array<Klass*>*>* a4 = ...;  it->push(&a3);    => MSOPointerArrayRef
```
Comment (calvinccheung) [inline src/hotspot/share/memory/metaspaceClosure.hpp:323]: &a3 should be &a4?

File: src/hotspot/share/memory/metaspaceClosure.hpp [hunk 5]
```diff
@@ -288,31 +306,69 @@ class MetaspaceClosure {
...
+  // Array<Annotation>*     a5 = ...;  it->push(&a3);    => MSOPointerArrayRef
```
Comment (calvinccheung) [inline src/hotspot/share/memory/metaspaceClosure.hpp:324]: Array<Annotation>* should be Array<Annotation*>* ?

&a3 should be &a5?

**Review by calvinccheung — APPROVED** [review]
Looks good. I spotted some typos in metaspaceClosure.hpp.

**Comment by iklam** [comment]
Thanks @fparain and @calvinccheung for the review!
/integrate

**Pull request closed without merging**